### ISSUE TITLE
Targeted Attack

### ISSUE THREAD

**Issue by postmachines** [locuslab/projected_sinkhorn#3]
Hello, thank you for your research! 

How I can apply targeted attack with projected Sinkhorn? 
It seems like only an untargeted attack available in code.

**Comment by riceric22** [comment]
Hi @postmachines, 

The projected Sinkhorn algorithm is non-specific to the adversarial loss function, since the core of the functionality is to project onto a Wasserstein ball (e.g. within a PGD loop as a subroutine). The PGD attack can maximize any loss (e.g. targeted or untargeted loss), and so to apply a targeted attack, you can simply change the untargeted loss to a targeted loss. 

For a concrete example, in the code you can see an example of the PGD attack at Line 29 of `pgd.py` at https://github.com/locuslab/projected_sinkhorn/blob/master/pgd.py#L29 which uses the cross entropy loss as an untargeted loss function. To make this a targeted attack, replace this with your desired targeted loss function (e.g. the logit of the correct class minus the logit of the target class). 

~Eric

**Comment by postmachines** [comment]
Thank you for your answer!

Sorry that my question is not issue related, but I still don’t understand how I can apply targeted adversarial attacks via maximizing any loss. 

Let me describe my question in more detail.   
In a standard PGD attack realization, to turn attack from untargeted into targeted, we just try to minimize the loss function instead of maximization https://github.com/MadryLab/robustness/blob/master/robustness/attacker.py#L139.

For example, In your code, to make a targeted attack using _linfinity_, I should just replace the line  https://github.com/locuslab/projected_sinkhorn/blob/master/pgd.py#L36 with `X_[~err] -= alpha*torch.sign(X_.grad[~err])`. 

But with _wasserstein_ it's not so clear, the conjugate_sinkhorn function https://github.com/locuslab/projected_sinkhorn/blob/master/pgd.py#L41 looks very complicated for me, and I have no idea how I can turn it into targeted version. The question is: Is any easy way to turn from maximization into minimization in your code? If not, any advice on how I can modify it will be very useful. 

~Arip.

**Comment by riceric22** [comment]
I see what you're talking about, in the MadryLab repository they implement a targeted attack by simply doing gradient ascent on the standard cross entropy loss with the true label replaced by the target label. Sorry for the confusion, as there are multiple ways to do a targeted attack (e.g. by optimizing a difference of logits instead). 

To change the PGD to an ascent, you're looking at the right spot. If you want to take the steepest descent step with respect to Wasserstein metric, you can just replace the `X_.grad` with `-X_.grad` in the conjugate Wasserstein function. This will flip the sign of the conjugate problem, and thus change the steepest ascent to steepest descent. 

However, I would recommend you take L2 or Linfinity steepest descent steps instead (and project it onto the Wasserstein ball). These tend to be empirically much more effective than the Wasserstein steepest descent step, which I've confirmed in discussions with others (e.g. faster computation and convergence). This is reflected in the default argument for the PGD function. 

Feel free to let me know if you have any other questions! 

~Eric
